### PR TITLE
CURA-3529 Copy the latest cura dir (if present) for upgrade

### PR DESCRIPTION
**Pull request by LipuFei**
In combination with https://github.com/Ultimaker/Cura/pull/1672

### CODE REVIEW

**Comment by fieldOfView** [comment]
Cura workarounds in Uranium? Hath hell frozen over? ;-)
This should probably be in an VersionUpgrade plugin.

**Comment by LipuFei** [comment]
@fieldOfView discussed with @nallath about this. If there is an earlier cura directory, we want to duplicate that one as the start point for upgrading to the newer version. This way the old configs are not touched and running the previous version will still work.

This needs to be done before we start anything, so we put the code here. `VersionUpgrade` plugin will just take care of upgrading the actual files.

**Comment by nallath** [comment]
We could make it a bit more generic by changing 
 "findLatestCuraDirInPaths" to "findLatestDirInPaths", which uses a more generic approach. 
As far as I can see, this should be doable.

**Comment by fieldOfView** [comment]
@LipuFei: I understand what this code is doing, it is just that all the references to Cura (and even a specific hardcoded version of Cura) inside a Uranium file are quite "illegal".

**Comment by LipuFei** [comment]
@fieldOfView @nallath Ah, right. Good point! I forgot we are in Uranium... I will change those too.

**Comment by LipuFei** [comment]
@fieldOfView @nallath code is more generic now, and some unit tests are added.

**Comment by fieldOfView** [comment]
I'm still not a fan of ```_isConfigDirForAppBefore26```. Version 2.6 of what?
Perhaps name it ```_isNonVersionedConfigDir```?

File: UM/Resources.py [hunk 1]
```diff
@@ -264,55 +268,188 @@ def __find(cls, resource_type: int, *args) -> List[str]:
...
-    def __initializeStoragePaths(cls):
-        # use nested structure: cura/<version>/...
-        if cls.ApplicationVersion == "master" or cls.ApplicationVersion == "unknown":
-            storage_dir_name = os.path.join(cls.ApplicationIdentifier, cls.ApplicationVersion)
-        else:
-            from UM.Version import Version
-            version = Version(cls.ApplicationVersion)
-            storage_dir_name = os.path.join(cls.ApplicationIdentifier, "%s.%s" % (version.getMajor(), version.getMinor()))
-
+    def getConfigStorageRootPath(cls):
```
Comment (awhiemstra) [inline UM/Resources.py]: This exposes an abbreviation as public API. (Config vs. Configuration) We generally try not use abbreviations, and should definitely not expose them as public API.

File: UM/Resources.py [hunk 2]
```diff
@@ -264,55 +268,188 @@ def __find(cls, resource_type: int, *args) -> List[str]:
...
-            cls.__config_storage_path = os.path.join(xdg_config_home, storage_dir_name)
+                config_path = os.path.expanduser("~/.config")
+        else:
+            config_path = "."
 
-            xdg_data_home = ""
+        return config_path
+
+    @classmethod
+    def getPossibleConfigStorageRootPathList(cls):
```
Comment (awhiemstra) [inline UM/Resources.py]: Why is this here? Can we move this to VersionUpgrade?

File: UM/Resources.py [hunk 3]
```diff
@@ -264,55 +268,188 @@ def __find(cls, resource_type: int, *args) -> List[str]:
...
+        # Returns all possible root paths for storing app configurations (in old and new versions)
+        config_root_list = [Resources.getConfigStorageRootPath()]
+        if platform.system() == "Windows":
```
Comment (awhiemstra) [inline UM/Resources.py]: Use UM.Platform.isWindows() instead.

File: UM/Resources.py [hunk 4]
```diff
@@ -264,55 +268,188 @@ def __find(cls, resource_type: int, *args) -> List[str]:
...
+            # it used to be in LOCALAPPDATA on Windows
+            config_root_list.append(os.getenv("LOCALAPPDATA"))
+
```
Comment (awhiemstra) [inline UM/Resources.py]: This misses the old storage location on OSX, which used to be ~/.cura . I cannot recall when we made that change though, so I do not know if we want to keep it.

File: UM/Resources.py [hunk 5]
```diff
@@ -264,55 +268,188 @@ def __find(cls, resource_type: int, *args) -> List[str]:
...
+        config_root_list = [os.path.join(n, cls.ApplicationIdentifier) for n in config_root_list]
+        return config_root_list
+
+    @classmethod
+    def getDataStorageRootPath(cls):
```
Comment (awhiemstra) [inline UM/Resources.py]: I would prefer it if we change these three to a "getStorageRootPath" that takes an enum to get the path we want. This makes the API more forward compatible and cleaner since adding an extra storage root just means adding an enum entry.

File: UM/Resources.py [hunk 6]
```diff
@@ -264,55 +268,188 @@ def __find(cls, resource_type: int, *args) -> List[str]:
...
 
-            xdg_cache_home = ""
+    @classmethod
+    def getCacheStorageRootPath(cls):
+        # Returns the path where we store different versions of app configurations
+        cache_path = None
+        if platform.system() == "Windows":
+            cache_path = os.getenv("LOCALAPPDATA")
+        elif platform.system() == "Darwin":
+            cache_path = None
```
Comment (awhiemstra) [inline UM/Resources.py]: Ehm, what? So getting the cache storage path for OSX leads to an invalid path?

File: UM/Resources.py [hunk 7]
```diff
@@ -264,55 +268,188 @@ def __find(cls, resource_type: int, *args) -> List[str]:
...
             except KeyError:
-                xdg_data_home = os.path.expanduser("~/.local/share")
-            cls.__data_storage_path = os.path.join(xdg_data_home, storage_dir_name)
+                data_path = os.path.expanduser("~/.local/share")
+        return data_path
 
-            xdg_cache_home = ""
+    @classmethod
+    def getCacheStorageRootPath(cls):
+        # Returns the path where we store different versions of app configurations
```
Comment (awhiemstra) [inline UM/Resources.py]: Documentation should be above the method declaration. Also, you are missing documentation for return type. I tend to use the following template:

```
##  Brief description of the method/class/etc.
#
#   Longer description of the item.
#
#   \param paramName \type{type} Description of the parameter. (Repeat for all parameters).
#
#   \return \type{type} Description of the return type.

File: UM/Resources.py [hunk 8]
```diff
@@ -264,55 +268,188 @@ def __find(cls, resource_type: int, *args) -> List[str]:
...
+            storage_dir_name = os.path.join(cls.ApplicationIdentifier, "%s.%s" % (version.getMajor(), version.getMinor()))
+
+        # config is saved in "<CONFIG_ROOT>/<storage_dir_name>"
+        cls.__config_storage_path = os.path.join(Resources.getConfigStorageRootPath(), storage_dir_name)
+
+        # data is saved in
+        #  - on Linux: "<DATA_ROOT>/<storage_dir_name>"
+        #  - on other: "<CONFIG_DIR>" (in the config directory)
+        data_root_path = Resources.getDataStorageRootPath()
+        cls.__data_storage_path = cls.__config_storage_path if data_root_path is None else \
```
Comment (awhiemstra) [inline UM/Resources.py]: Please just write this as

```
if data_root_path is None:
  cls.__data_storage_path = cls.__config_storage_path
else:
  cls.__data_storage_path = os.path.join(data_root_path, storage_dir_name)
```

This prevents ugly escaping of newlines and is generally much more readable.

File: UM/Resources.py [hunk 9]
```diff
@@ -264,55 +268,188 @@ def __find(cls, resource_type: int, *args) -> List[str]:
...
+
+        if not os.path.exists(cls.__config_storage_path):
+            cls._copyLatestDirsIfPresent()
 
         cls.__paths.insert(0, cls.__data_storage_path)
 
+    ##  Copies the directories of the latest version on this machine if present, so the upgrade will use the copies
+    #   as the base for upgrade
+    @classmethod
+    def _copyLatestDirsIfPresent(cls):
```
Comment (awhiemstra) [inline UM/Resources.py]: Why is this here? Can we move this to VersionUpgrader?

File: UM/Resources.py [hunk 10]
```diff
@@ -264,55 +268,188 @@ def __find(cls, resource_type: int, *args) -> List[str]:
...
+                for dir_name in storage_dir_name_list:
+                    if not dir_name.endswith("master"):
+                        latest_storage_dir_path = os.path.join(search_path, dir_name)
+                        break
+            if latest_storage_dir_path is not None:
+                break
+        return latest_storage_dir_path
+
+    @classmethod
+    def _isConfigDirForAppBefore26(cls, check_path):
```
Comment (awhiemstra) [inline UM/Resources.py]: I agree with Aldo that this should have a different name.

File: UM/Resources.py [hunk 11]
```diff
@@ -264,55 +268,188 @@ def __find(cls, resource_type: int, *args) -> List[str]:
...
+            return True
+
+        dirs, files = next(os.walk(check_path))[1:]
+        valid_dir_names = [dn for dn in dirs if dn in Resources.__expected_dir_names_in_config]
+        valid_file_names = [fn for fn in files if not fn.endswith('.log')]  # do not include log files
+
+        return valid_dir_names and valid_file_names
+
+    @classmethod
+    def addExpectedDirNameInConfig(cls, dir_name):
```
Comment (awhiemstra) [inline UM/Resources.py]: What does this do?

File: UM/Resources.py [hunk 12]
```diff
@@ -264,55 +268,188 @@ def __find(cls, resource_type: int, *args) -> List[str]:
...
+        if not latest_config_path:
+            # no earlier storage dirs found, do nothing
+            return
+
+        if latest_config_path == this_version_config_path:
+            # if the directory found matches the current version, do nothing
+            return
+
+        # we first copy everything to a temporary folder, and then move it to the new folder
+        def _copyDirToDirViaTempDir(src_path, dest_path):
```
Comment (awhiemstra) [inline UM/Resources.py]: Does this need to be a closure?

**Comment by LipuFei** [comment]
@fieldOfView That's a better name. I will fix it.

File: UM/Resources.py (hunk 12 above)
Comment (LipuFei) [inline UM/Resources.py]: The closure can go. It just means that this function is not meant for anything out of this scope.

File: UM/Resources.py (hunk 9 above)
Comment (LipuFei) [inline UM/Resources.py]: The VersionUpgrader is an plugin, which means the plugins will already he loaded when VersionUpgrader is launched. A user can have custom plugins in his/her old app config directory, e.g. `cura/plugins`. If the VersionUpgrader decides to upgrade from that old version, Cura needs to restart to load those custom plugins. That's why we decide to run those code before actually doing anything.

File: UM/Resources.py (hunk 8 above)
Comment (LipuFei) [inline UM/Resources.py]: Will do

File: UM/Resources.py (hunk 6 above)
Comment (LipuFei) [inline UM/Resources.py]: There are some fuzzy logic in the old code too. Basically, there must be a path for the **config dir**. If no place can be found for it, it will be `.` the current directory. But for **data** and **cache**, it depends. Only on Linux do we have separate places for all of them; On Windows **config** and **data** are kept in the Roaming, and **cache** in Local; On Mac all of them are in the same place. This basically means **data** and **cache** are optional to have their own places, thus the `None` in `getCache` and `getData`.

File: UM/Resources.py (hunk 4 above)
Comment (LipuFei) [inline UM/Resources.py]: Will fix

File: UM/Resources.py (hunk 11 above)
Comment (LipuFei) [inline UM/Resources.py]: This is to detect an unversioned directory.

File: UM/Resources.py (hunk 6 above)
Comment (awhiemstra) [inline UM/Resources.py]: Yeah, but the old code ensured we always had a proper path, even if that path might be the same as another. Right now, what this code does is lead to a call like "os.path.join(Resources.getCacheRootPath(), "blah") ending up with either a TypeError or an invalid path.
